### PR TITLE
fix: uninstall jenkins-x

### LINKED ISSUE #7165: Unable to `jx uninstall`
`jx uninstall` hangs when running against EKS cluster. The problems seem to be that the deletion of the _PersistentVolumeClaims_ gets stuck in Terminating state.

```
? This action will permanently delete Jenkins X from the Kubernetes context foo. Please type in the name of the context to confirm: foo
Removing installation of Jenkins X in team namespace jx
WARNING: Not deleting jx-prow because the release is not installed

error: [failed to uninstall the jenkins-x helm chart in namespace jx: [failed to run 'kubectl delete pvc --ignore-not-found -l jenkins.io/chart-release=jenkins-x --namespace jx --wait' command in directory '', output: 'persistentvolumeclaim "jenkins" deleted
persistentvolumeclaim "jenkins-x-chartmuseum" deleted
persistentvolumeclaim "jenkins-x-docker-registry" deleted
persistentvolumeclaim "jenkins-x-nexus" deleted
```

Manually running `kubectl delete pvc --ignore-not-found -l jenkins.io/chart-release=jenkins-x --namespace jx` also hangs.

Could affect GKE as well. Only verified against EKS so far.

### PR DESCRIPTION
**Pull request by mrageh**
#### Submitter checklist

- [x] Change is code complete and matches issue description.
- [ ] Change is covered by existing or new tests.

#### Description
When uninstalling `jx`, the PVC would hang in a terminating state. When we set the PVC, finalizer volume to `null`, the uninstall process completes as normal.

There is an open issue in [Kubernetes](https://github.com/kubernetes/kubernetes/issues/77258) that has not yet been fixed.

This PR sets the PVC finalizer to `null`, which fixes our issue and allows users to uninstall `jx`.

#### Which issue this PR fixes

fixes #7165

### CODE REVIEW

File: pkg/cmd/uninstall/uninstall.go [hunk 1]
```diff
@@ -245,3 +252,19 @@ func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string,
...
 	return errors
 }
+
+func (o *UninstallOptions) setPVCFinalizerToNull(ns string, releaseName string) error {
+	client, err := o.KubeClient()
+	if err != nil {
+		return errors.Wrap(err, "getting the kube client")
+	}
+
+	list, _ := client.CoreV1().PersistentVolumeClaims(ns).List(meta_v1.ListOptions{})
```
Comment (dgozalo) [inline pkg/cmd/uninstall/uninstall.go:262]: I think we should not ignore this error, there may be a problem with the connection there. The list will be empty instead of nil so there won't be a panic, but I think we should handle all errors.

File: pkg/cmd/uninstall/uninstall.go [hunk 2]
```diff
@@ -245,3 +252,19 @@ func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string,
...
+
+	for _, pvc := range list.Items {
+		pvc.Finalizers = []string{}
+		client.CoreV1().PersistentVolumeClaims(ns).Update(&pvc)
```
Comment (dgozalo) [inline pkg/cmd/uninstall/uninstall.go:266]: Same here, we should be handling the error on update.

File: pkg/cmd/uninstall/uninstall.go [hunk 3]
```diff
@@ -236,6 +236,13 @@ func (o *UninstallOptions) deleteNamespace(namespace string) error {
 // of an unrelated error
 func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string, errors []error, force bool) []error {
 	if err := o.Helm().StatusRelease(ns, releaseName); err == nil || force {
+
+		err = o.setPVCFinalizerToNull(ns, releaseName)
+
+		if err != nil {
+			errors = append(errors, fmt.Errorf("failed to uninstall the %s helm chart in namespace %s: %s", releaseName, ns, err))
```
Comment (dgozalo) [inline pkg/cmd/uninstall/uninstall.go:243]: This error actually comes from the `setPVCFinalizerToNull`, the error message should be related.

File: pkg/cmd/uninstall/uninstall.go [hunk 4]
```diff
@@ -236,6 +236,13 @@ func (o *UninstallOptions) deleteNamespace(namespace string) error {
 // of an unrelated error
 func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string, errors []error, force bool) []error {
 	if err := o.Helm().StatusRelease(ns, releaseName); err == nil || force {
+
+		err = o.setPVCFinalizerToNull(ns, releaseName)
+
+		if err != nil {
+			errors = append(errors, fmt.Errorf("Failed to set PVC finalizer to null in %s helm chart in namespace %s: %s", releaseName, ns, err))
```
Comment (ankitm123) [inline pkg/cmd/uninstall/uninstall.go:243]: Any reason we don't return when we see an error? curious ...
More specifically, should we even go to the next step (in this case DeleteRelease) if an error is encountered?

Comment (mrageh) [reply pkg/cmd/uninstall/uninstall.go:243]: Good point, let me fix that and return immediately!

**Comment by hferentschik** [comment]
from the unit test logs:
```
# github.com/jenkins-x/**/v2/pkg/cmd/uninstall
pkg/cmd/uninstall/uninstall.go:274:11: Wrapf format %s has arg pvc of wrong type k8s.io/api/core/v1.PersistentVolumeClaim
```

File: pkg/cmd/uninstall/uninstall.go (hunk 4 above)
Comment (mrageh) [reply pkg/cmd/uninstall/uninstall.go:243]: Fixed!

**Comment by mrageh** [comment]
Fixed, this is ready to merge

**Comment by dgozalo** [comment]
I think this could use a unit test. Just need to create a few PVCs using the mock and test that your code actually sets the Finalizers to nil.

File: pkg/cmd/uninstall/uninstall.go [hunk 5]
```diff
@@ -236,6 +236,14 @@ func (o *UninstallOptions) deleteNamespace(namespace string) error {
 // of an unrelated error
 func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string, errors []error, force bool) []error {
 	if err := o.Helm().StatusRelease(ns, releaseName); err == nil || force {
+
+		err = o.setPVCFinalizerToNull(ns, releaseName)
```
Comment (hferentschik) [inline pkg/cmd/uninstall/uninstall.go:240]: Actually, out of interest, why do you call `setPVCFinalizerToNull` from here? The way `setPVCFinalizerToNull` is implemented, it is a global thing w/o any dependence to the release name.

File: pkg/cmd/uninstall/uninstall.go [hunk 6]
```diff
@@ -245,3 +253,27 @@ func (o *UninstallOptions) DeleteReleaseIfPresent(ns string, releaseName string,
 	}
 	return errors
 }
+
+func (o *UninstallOptions) setPVCFinalizerToNull(ns string, releaseName string) error {
```
Comment (hferentschik) [inline pkg/cmd/uninstall/uninstall.go:257]: _releaseName_ is not really used. I think `setPVCFinalizerToNull` is really a "global" function to be executed somewhere from within `Run`.

**Comment by hferentschik** [comment]
@mrageh, @dgozalo - fyi, I pushed another commit to this pull request. It removes the unused parameter from `setPVCFinalizerToNull` and moves the function call itself. 

Also adding `jx uninstall` as a task to the _tf-boot_ context cleanup. This way we get a test of it as well.
